operator/pkg/ingress/ingress.go

```diff
@@ -5,720 +5,300 @@ package ingress
-	if !cache.WaitForCacheSync(wait.NeverStop, ic.ingressInformer.HasSynced) {
-		return fmt.Errorf("unable to wait for Ingress cache sync")
-	}
-
-	go ic.serviceManager.Run()
-
-	for ic.processEvent() {
-	}
-
-	return nil
-}
-
-func (ic *Controller) processEvent() bool {
-	event, shutdown := ic.queue.Get()
-	if shutdown {
-		return false
-	}
-	defer ic.queue.Done(event)
-	err := ic.handleEvent(event)
-	if err == nil {
-		ic.queue.Forget(event)
-	} else if ic.queue.NumRequeues(event) < ic.maxRetries {
-		ic.queue.AddRateLimited(event)
-	} else {
-		log.WithError(err).Errorf("Failed to process Ingress event, skipping: %s", event)
-		ic.queue.Forget(event)
-	}
-	return true
+	sharedTranslator    translation.Translator
+	dedicatedTranslator translation.Translator
 }
 
-func getIngressClassName(ingress *slim_networkingv1.Ingress) *string {
-	annotations := ingress.GetAnnotations()
-	if className, ok := annotations["kubernetes.io/ingress.class"]; ok {
-		return &className
+func newIngressReconciler(
+	logger logrus.FieldLogger,
+	c client.Client,
+	ciliumNamespace string,
+	enforceHTTPS bool,
+	useProxyProtocol bool,
+	secretsNamespace string,
+	lbAnnotationPrefixes []string,
+	sharedLBServiceName string,
+	defaultLoadbalancerMode string,
+	defaultSecretNamespace string,
+	defaultSecretName string,
+	proxyIdleTimeoutSeconds int,
+) *ingressReconciler {
+	return &ingressReconciler{
+		logger: logger,
+		client: c,
+
+		sharedTranslator:    ingressTranslation.NewSharedIngressTranslator(sharedLBServiceName, ciliumNamespace, secretsNamespace, enforceHTTPS, useProxyProtocol, proxyIdleTimeoutSeconds),
+		dedicatedTranslator: ingressTranslation.NewDedicatedIngressTranslator(secretsNamespace, enforceHTTPS, useProxyProtocol, proxyIdleTimeoutSeconds),
+
+		maxRetries:              10,
+		enforcedHTTPS:           enforceHTTPS,
+		useProxyProtocol:        useProxyProtocol,
+		secretsNamespace:        secretsNamespace,
+		lbAnnotationPrefixes:    lbAnnotationPrefixes,
+		sharedLBServiceName:     sharedLBServiceName,
+		ciliumNamespace:         ciliumNamespace,
+		defaultLoadbalancerMode: defaultLoadbalancerMode,
+		defaultSecretNamespace:  defaultSecretNamespace,
+		defaultSecretName:       defaultSecretName,
+		idleTimeoutSeconds:      proxyIdleTimeoutSeconds,
 	}
-
-	return ingress.Spec.IngressClassName
 }
 
-func hasEmptyIngressClass(ingress *slim_networkingv1.Ingress) bool {
-	className := getIngressClassName(ingress)
-
-	return className == nil || *className == ""
-}
-
-func (ic *Controller) isCiliumIngressEntry(ingress *slim_networkingv1.Ingress) bool {
-	className := getIngressClassName(ingress)
-
-	if (className == nil || *className == "") && ic.ingressClassManager.IsDefault() {
-		return true
-	}
-
-	return className != nil && *className == ciliumIngressClassName
+// SetupWithManager sets up the controller with the Manager.
+func (r *ingressReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
```

Comment:
the current implementation didn't watch backend services, WDYT about adding such feature in?

Basically, it will help for the case that backend service is created _after_ Ingress creation.

Reply:
do we actually rely on data from (and therefore load) the backend service(s) that is referenced by the Ingress? :thinking: 

If yes, we need a watch.
It not, it should not be necessary.

Reply:
you are right, it's not required as CEC will take the service name and perform the needful.